### PR TITLE
[ROCm] Optimized training apply ops

### PR DESCRIPTION
**Pull request by ekuznetsov139**
This PR implements GPU kernels ApplyAdagrad, ApplyAdagradV2, ApplyAdadelta, ApplyRMSProp, and ApplyCenteredRMSProp for ROCm, and enables corresponding unit tests.

### CODE REVIEW

File: tensorflow/core/kernels/training_ops.cc [hunk 1]
```diff
@@ -670,9 +670,8 @@ namespace functor {
 DECLARE_GPU_SPEC(Eigen::half);
 DECLARE_GPU_SPEC(float);
 DECLARE_GPU_SPEC(double);
-#if !defined(TENSORFLOW_USE_NVCC) && \
-    !defined(TENSORFLOW_USE_ROCM)  // TODO(b/143684500): Eigen to support
-                                   // complex sqrt
+#if !defined( \
```
Comment (chsigg) [inline tensorflow/core/kernels/training_ops.cc]: `#ifndef` for consistency.

**Comment by gbaned** [comment]
@ekuznetsov139 Can you please resolve conflicts? Thanks!

**Comment by deven-amd** [comment]
gentle ping

**Comment by tanzhenyu** [comment]
I'm not very familiar with ROCM implementations, @chsigg might know better.

**Comment by deven-amd** [comment]
gentle ping

File: tensorflow/core/kernels/training_ops_gpu.cu.cc [hunk 2]
```diff
@@ -118,19 +118,192 @@ struct ApplyGradientDescent<GPUDevice, T> {
...
+    ms[i] += one_minus_rho * (grad[i] * grad[i] - ms[i]);
+    mg[i] += one_minus_rho * (grad[i] - mg[i]);
+    T denom = (ms[i] - mg[i] * mg[i]) + eps;
+    mom[i] = mom[i] * momentum + lr * grad[i] * impl_rsqrt(denom);
+    var[i] -= mom[i];
+  }
+}
+#endif
+
+#if TENSORFLOW_USE_ROCM
```
Comment (cheshire) [inline tensorflow/core/kernels/training_ops_gpu.cu.cc]: Nitpick: why not remove this `#if` and the previous `#endif`?

File: tensorflow/core/kernels/training_ops_gpu.cu.cc [hunk 3]
```diff
@@ -118,19 +118,192 @@ struct ApplyGradientDescent<GPUDevice, T> {
...
+__device__ Eigen::half impl_rsqrt(Eigen::half x) {
+  return __float2half(rsqrt(__half2float(x)));
+}
+
+template <class T>
+__device__ std::complex<T> impl_sqrt(std::complex<T> x) {
+  T re = x.real(), im = x.imag();
+  T mod_x = sqrt(re * re + im * im);
+  const T root2 = 0.7071067811865475;
+  // we pick the root with the same sign of the imaginary component as the input
```
Comment (cheshire) [inline tensorflow/core/kernels/training_ops_gpu.cu.cc]: Nitpick: comments have to be full sentences, start with a capital letter, and end with a full stop.

**Review by cheshire — DISMISSED** [review]
LGTM with minor nits.

**Comment by gbaned** [comment]
@ekuznetsov139 Can you please check @cheshire's comments and keep us posted. Thanks!

**Review by cheshire — APPROVED** [review]

**Pull request merged**